### PR TITLE
Feature/modernize cmake

### PR DESCRIPTION
**Pull request by ClausKlein**
according suggestions form [cmake discourse](https://discourse.cmake.org/t/is-this-really-a-better-style-to-use-generator-expressions-every-time/2753) and

https://crascit.com/2019/10/16/cppcon-2019-deep-cmake-for-library-authors/

and my experience

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -63,14 +95,23 @@ target_include_directories(
...
 # header paths
 string(TOLOWER ${PROJECT_NAME}/version.h VERSION_HEADER_LOCATION)
+string(TOLOWER ${PROJECT_NAME}/greeter_export.h EXPORT_HEADER_LOCATION)
+
+# TODO: should be done in packageProject()
+include(GenerateExportHeader)
+generate_export_header(
+  ${PROJECT_NAME} EXPORT_FILE_NAME PackageProjectInclude/${EXPORT_HEADER_LOCATION}
+)
+# Note: the export header will be installed while installing the version header! CK
```
Comment (ClausKlein) [inline CMakeLists.txt:105]: do you mean, this would be a possible extension to `packageProject.cmake`?

File: CMakeLists.txt [hunk 2]
```diff
@@ -63,14 +95,23 @@ target_include_directories(
...
   NAME ${PROJECT_NAME}
   VERSION ${PROJECT_VERSION}
   NAMESPACE ${PROJECT_NAME}
   BINARY_DIR ${PROJECT_BINARY_DIR}
   INCLUDE_DIR ${PROJECT_SOURCE_DIR}/include
-  INCLUDE_DESTINATION include/${PROJECT_NAME}-${PROJECT_VERSION}
+  INCLUDE_DESTINATION ${INCLUDE_INSTALL_DIR}
   VERSION_HEADER "${VERSION_HEADER_LOCATION}"
-  DEPENDENCIES ""
+  COMPATIBILITY SameMajorVersion
```
Comment (ClausKlein) [inline CMakeLists.txt:115]: this would be the recomented upgrade strategie

File: include/greeter/greeter.h [hunk 3]
```diff
@@ -1,31 +1,33 @@
...
 
   /**  Language codes to be used with the Greeter class */
-  enum class LanguageCode { EN, DE, ES, FR };
+  enum class GREETER_EXPORT LanguageCode { EN, DE, ES, FR };
 
   /**
    * @brief A class for saying hello in multiple languages
    */
-  class Greeter {
+  class GREETER_EXPORT Greeter {
```
Comment (ClausKlein) [inline include/greeter/greeter.h:15]: Do you have any idea, why `MSC` requires to export the private member `name`?

File: test/CMakeLists.txt [hunk 4]
```diff
@@ -14,15 +18,16 @@ include(../cmake/tools.cmake)
...
 CPMAddPackage(
   NAME doctest
   GITHUB_REPOSITORY onqtam/doctest
-  GIT_TAG 2.3.7
+  GIT_TAG 2.4.5
 )
 
 if(TEST_INSTALLED_VERSION)
-  find_package(Greeter REQUIRED)
+  find_package(Greeter ${PROJECT_VERSION} REQUIRED)
```
Comment (ClausKlein) [inline test/CMakeLists.txt:30]: Note: this match the upgrade strategy!

File: test/source/greeter.cpp [hunk 5]
```diff
@@ -16,6 +14,11 @@ TEST_CASE("Greeter") {
 }
 
 TEST_CASE("Greeter version") {
-  static_assert(std::string_view(GREETER_VERSION) == std::string_view("1.0"));
-  CHECK(std::string(GREETER_VERSION) == std::string("1.0"));
-}
\ No newline at end of file
+#if (__APPLE__ || __cpp_lib_starts_ends_with)
+  static_assert(std::string_view(GREETER_VERSION).starts_with("1"));  // TBD C++20 only
+  CHECK(std::string(GREETER_VERSION).starts_with("1"));               // SameMajorVersion
```
Comment (ClausKlein) [inline test/source/greeter.cpp:19]: Note: this match the upgrade strategy too!

**Comment by bhardwajs** [comment]
@ClausKlein - What problem are you running into in building DLL on windows using CMake? I use CMake and have built DLLs and can help here.

File: standalone/CMakeLists.txt [hunk 6]
```diff
@@ -21,10 +22,17 @@ CPMAddPackage(NAME Greeter SOURCE_DIR ${CMAKE_CURRENT_LIST_DIR}/..)
...
+set_target_properties(
+  GreeterStandalone PROPERTIES CXX_STANDARD ${CMAKE_CXX_STANDARD} OUTPUT_NAME "Greeter"
+)
 
-target_link_libraries(GreeterStandalone Greeter::Greeter cxxopts)
+# WORKAROUND missing ALIAS target error! CK
+if(NOT TARGET cxxopts::cxxopts)
+  add_library(cxxopts::cxxopts ALIAS cxxopts)
+endif()
+target_link_libraries(GreeterStandalone Greeter::Greeter cxxopts::cxxopts)
```
Comment (ClausKlein) [inline standalone/CMakeLists.txt:38]: Note: the cxxopts is not clean build, missing alias!

File: cmake/tools.cmake [hunk 7]
```diff
@@ -7,6 +7,23 @@ if(NOT PROJECT_SOURCE_DIR STREQUAL CMAKE_SOURCE_DIR)
 endif()
 
 include(${CMAKE_CURRENT_LIST_DIR}/CPM.cmake)
+CPMUsePackageLock(package-lock.cmake)
+
+set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
+set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
+set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
```
Comment (ClausKlein) [inline cmake/tools.cmake:14]: Note: needed to build executable and dll at same directory on windows!

File: .github/workflows/windows.yml [hunk 8]
```diff
@@ -25,12 +28,12 @@ jobs:
           key: ${{ github.workflow }}-cpm-modules-${{ hashFiles('**/CMakeLists.txt', '**/*.cmake') }}
 
       - name: configure
-        run: cmake -Stest -Bbuild
+        run: cmake -S test -B build   # XXX -G "Visual Studio 16 2019" -T ClangCL
```
Comment (ClausKlein) [inline .github/workflows/windows.yml:31]: Note: the `ClangCL` does not work yet! I have no idea why?

**Review by ClausKlein — COMMENTED** [review]
> @ClausKlein - What problem are you running into in building DLL on windows using CMake? I use CMake and have built DLLs and can help here.

Thank you, luckily I found a solution. Not `building`, running the `executable` was the Problem.

**Pull request closed without merging**